### PR TITLE
MODINVOICE-17 Define API

### PR DESCRIPTION
**Pull request by javalkar-varun**
## Purpose

https://issues.folio.org/browse/MODINVOICE-17

## Approach

- Update invoice.raml with all APIs
- Update ModuleDescriptor with appropriate permissions
- Generate all interfaces for Invoices implementation
- Update `.gitmodules` to create ramls/acq-models sub-folder
- Update unit tests to comply with the new interface

## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [x] Code coverage on new code is 80% or greater
  - [x] Duplications on new code is 3% or less
  - [x] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [x] There are no breaking changes in this PR.
  
If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?  
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.  

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

File: descriptors/ModuleDescriptor-template.json [hunk 1]
```diff
@@ -9,29 +9,64 @@
       "handlers": [
         {
           "methods": ["GET"],
-          "pathPattern": "/invoice",
-          "permissionsRequired": ["invoice.collection.get"]
+          "pathPattern": "/invoicing/invoices",
+          "permissionsRequired": ["invoicing.collection.get"]
```
Comment (craigmcnally) [inline descriptors/ModuleDescriptor-template.json]: invoice.collection.get

Comment (craigmcnally) [reply descriptors/ModuleDescriptor-template.json]: the first is probably more consistent with the other perms here

File: descriptors/ModuleDescriptor-template.json [hunk 2]
```diff
@@ -9,29 +9,64 @@
...
         },
         {
           "methods": ["POST"],
-          "pathPattern": "/invoice",
-          "permissionsRequired": ["invoice.item.post"]
+          "pathPattern": "/invoicing/invoices",
+          "permissionsRequired": ["invoicing.item.post"]
```
Comment (craigmcnally) [inline descriptors/ModuleDescriptor-template.json]: see above comments

**Comment by craigmcnally** [comment]
Whoops, looks like there are some raml-cop issues you need to resolve

**Comment by javalkar-varun** [comment]
@craigmcnally Done, I had an extra copy of schema locally which was throwing an error. Removing those fixed raml-lint-cop error.

File: descriptors/ModuleDescriptor-template.json [hunk 3]
```diff
@@ -9,34 +9,69 @@
...
-          "pathPattern": "/invoice",
-          "permissionsRequired": ["invoice.collection.get"]
+          "pathPattern": "/invoicing/invoices",
+          "permissionsRequired": ["invoicing.invoice.collection.get"]
         },
         {
           "methods": ["POST"],
-          "pathPattern": "/invoice",
-          "permissionsRequired": ["invoice.item.post"]
+          "pathPattern": "/invoicing/invoices",
```
Comment (KVupp) [inline descriptors/ModuleDescriptor-template.json]: https://dev.folio.org/guidelines/naming-conventions/ . According to the naming conventions
>The first portion is the exact name of the responsible module (back-end modules drop the mod- prefix).

Should we be renaming the module? or the permissions

Comment (javalkar-varun) [reply descriptors/ModuleDescriptor-template.json]: If we don't rename the module then permission names will be like:
`invoice.invoice.item.post` 
`invoice.invoice-lines.item.post` ?

Comment (javalkar-varun) [reply descriptors/ModuleDescriptor-template.json]: I would prefer this:
`invoice.item.post`
`invoice-lines.item.post`
`invoice-number.item.post`
Any thoughts @craigmcnally @piotr-kalashuk ?

Comment (craigmcnally) [reply descriptors/ModuleDescriptor-template.json]: Damn it, I was unaware of the "**exact** name of the responsible module" part.  

I guess it should be:
`invoice.invoices.item.post` 
`invoice.invoices-lines.item.post`
etc.

Sorry for the confusion @javalkar-varun I should have known about this convention.  Good catch @KVupp

Comment (javalkar-varun) [reply descriptors/ModuleDescriptor-template.json]: No worries. I have corrected permission names.

File: ramls/invoice.raml [hunk 4]
```diff
@@ -23,30 +28,100 @@ traits:
...
+    displayName: Invoice Lines
+    description: Manage Invoice lines
+    type:
+      collection:
+        schemaCollection: invoiceLineCollection
+        schemaItem: invoiceLine
+        exampleCollection: !include acq-models/mod-invoice-storage/examples/invoice_line_collection.sample
+        exampleItem: !include acq-models/mod-invoice-storage/examples/invoice_line.sample
+    is: [validate]
+    get:
```
Comment (KVupp) [inline ramls/invoice.raml]: please add pageable to get end point

**Pull request merged**